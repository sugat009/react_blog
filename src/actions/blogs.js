import uuid from 'uuid';
import { database } from '../firebase/firebase';


// add blog post
export const addBlog = (blog) => ({
    type: 'ADD_BLOG',
    blog
});

// start add blog (async action for addBlog)
export const startAddBlog = (blogData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            content = '',
            createdAt = 0
        } = blogData;
        const blog = { title, content, createdAt };
        return database.ref(`users/${uid}/blogs`).push(blog)
            .then((ref) => {
                dispatch(addBlog({
                    id: ref.key,
                    ...blog
                }));
            });
    };
};

// remove blog post
export const removeBlog = ({ id } = {}) => ({
    type: 'REMOVE_BLOG',
    id
});

export const startRemoveBlog = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).remove().then(() => {
            dispatch(removeBlog({ id }));
        });
    };
};


// edit blog post
export const editBlog = (id, updates) => ({
    type: 'EDIT_BLOG',
    id,
    updates
});

export const startEditBlog = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).update(updates)
            .then(() => {
                dispatch(editBlog(id, updates));
            });
    };
};


// SET Blogs part
export const setBlog = (blogs) => ({
    type: 'SET_BLOG',
    blogs
});

export const setStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                blogData.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setBlog(blogData));
        });
    };  
};

// GET blogs part
export const getBlog = (blogs) => ({
    type: 'GET_BLOG',
    blogs
});

export const getStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                if (childSnapshot.key !== getState().auth.uid) {
                    blogData.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val().blogs
                    });
                }
            });
            dispatch(getBlog(blogData));
            console.log(blogData);
        });
    };
};