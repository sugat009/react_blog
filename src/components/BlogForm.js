import React from 'react';
import moment from 'moment';

export default class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.blog ? props.blog.title : '',
            content: props.blog ? props.blog.content : '',
            createdAt: props.blog ? moment(props.blog.createdAt) : moment(0),
            errors: ''
        };
    };
    contentChange = (e) => {
        const content = e.target.value;
        this.setState({
            content
        });
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState({
            title
        });
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.content) {
            this.setState({
                errors: "Both title and content are required!"
            });
        } else {
            this.setState(() => ({ errors: ''}));
            if (moment(this.state.createdAt).isSame(moment(0))) {
                this.props.onSubmit({
                    title: this.state.title,
                    content: this.state.content,
                    createdAt: moment().valueOf()
                });
            } else {
                this.props.onSubmit({
                    title: this.state.title,
                    content: this.state.content
                });
            }
        }
    };
    render() {
        return (
            <form onSubmit={this.onSubmit} className="form">
                {this.state.errors && <p className="form__error">{this.state.errors}</p> }
                <input 
                    className="text-input"
                    type="text"
                    placeholder="Blog Title"
                    autoFocus
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <textarea 
                    className="text-area"
                    placeholder="Blog Content"
                    value={this.state.content}
                    onChange={this.contentChange}
                />
                <div>
                    <button className="button">Save Blog</button>
                </div>
            </form>
        )
    };
};