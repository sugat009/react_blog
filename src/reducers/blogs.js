const blogReducerDefaultState =  [];

export default (state = blogReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BLOG':
            return [
                ...state,
                action.blog
            ];
        case 'REMOVE_BLOG':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_BLOG':
            return state.map((blog) => {
                if (blog.id === action.id) {
                    return {
                        ...blog,
                        ...action.updates
                    }
                } else {
                    return blog;
                };
            });
        case 'SET_BLOG':
            console.log(action.blogs, "aaaaaaaaaaaaaaaaaaaa");
            return action.blogs;
        case 'GET_BLOG':
            console.log(action.blogs);
            return action.blogs;
        default:
            return state;
    }
};