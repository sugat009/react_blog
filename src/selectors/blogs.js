import moment from 'moment';

// get visible blogs after the filters have been applied

export default (blogs, { title, sortBy, startDate, endDate }) => {
    return blogs.filter((blog) => {
        const createdAtMoment = moment(blog.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = blog.title.toLowerCase().includes(title.toLowerCase());
        
        return startDateMatch && endDateMatch && titleMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'alphabet') {
            return a.title > b.title ? 1 : -1;
        }
    });
};