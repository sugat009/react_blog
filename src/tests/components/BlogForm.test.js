import React from 'react'
import { shallow } from 'enzyme';
import moment from 'moment';
import BlogForm from '../../components/BlogForm';
import blogs from '../fixtures/blogs';


test('should render BlogForm correctly', () => {
    const wrapper = shallow(<BlogForm />);
    expect(wrapper).toMatchSnapshot();
});

// ExpenseForm with data
test('should render BlogForm with data', () => {
    const wrapper = shallow(<BlogForm blog={blogs[2]} />);
    expect(wrapper).toMatchSnapshot();
});

// error rendering
test('should render error', () => {
    const wrapper = shallow(<BlogForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('errors').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set title on input change', () => {
    const value = "some title"; // this variable must be named value(compulsory)
    const wrapper = shallow(<BlogForm />);
    wrapper.find('input').simulate('change', {
        target: { value }
    });
    expect(wrapper.state('title')).toBe(value);
});

test('should set content on textarea change', () => {
    const value = 'New content value';
    const wrapper = shallow(<BlogForm />);
    wrapper.find('textarea').simulate('change', {
      target: { value }
    });
    expect(wrapper.state('content')).toBe(value);
});

test('should call onSubmit prop for form submit', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<BlogForm blog={blogs[2]} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => { }
    });
    expect(wrapper.state('errors')).toBeFalsy();
    expect(onSubmitSpy).toHaveBeenCalledWith({ // blog object is expanded because id is also returned
        title: blogs[2].title,
        content: blogs[2].content,
        createdAt: blogs[2].createdAt
    });
});