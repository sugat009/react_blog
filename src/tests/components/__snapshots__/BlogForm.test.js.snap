// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Blog Title"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Blog Content"
      value=""
    />
    <button>
      Save Blog
    </button>
  </form>
</div>
`;

exports[`should render BlogForm with data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Blog Title"
      type="text"
      value="C# MasterClass"
    />
    <textarea
      onChange={[Function]}
      placeholder="Blog Content"
      value="Learn everything from scratch in C#"
    />
    <button>
      Save Blog
    </button>
  </form>
</div>
`;

exports[`should render error 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Blog Title"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Blog Content"
      value=""
    />
    <button>
      Save Blog
    </button>
  </form>
</div>
`;

exports[`should render error 2`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <p>
      Both title and content are required!
    </p>
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Blog Title"
      type="text"
      value=""
    />
    <textarea
      onChange={[Function]}
      placeholder="Blog Content"
      value=""
    />
    <button>
      Save Blog
    </button>
  </form>
</div>
`;
